src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RedirectHandler.cs

```diff
@@ -0,0 +1,126 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed partial class RedirectHandler : HttpMessageHandler
+    {
+        private readonly HttpMessageHandler _initialInnerHandler;       // Used for initial request
+        private readonly HttpMessageHandler _redirectInnerHandler;      // Used for redirects; this allows disabling auth
+        private readonly int _maxAutomaticRedirections;
+
+        public RedirectHandler(int maxAutomaticRedirections, HttpMessageHandler initialInnerHandler, HttpMessageHandler redirectInnerHandler)
+        {
+            Debug.Assert(initialInnerHandler != null);
+            Debug.Assert(redirectInnerHandler != null);
+
+            if (maxAutomaticRedirections < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(maxAutomaticRedirections));
+            }
+
+            _maxAutomaticRedirections = maxAutomaticRedirections;
+            _initialInnerHandler = initialInnerHandler;
+            _redirectInnerHandler = redirectInnerHandler;
+        }
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            HttpResponseMessage response = await _initialInnerHandler.SendAsync(request, cancellationToken).ConfigureAwait(false);
+
+            uint redirectCount = 0;
+            Uri redirectUri;
+            while ((redirectUri = GetUriForRedirect(request.RequestUri, response)) != null)
+            {
+                response.Dispose();
+
+                redirectCount++;
+                if (redirectCount > _maxAutomaticRedirections)
+                {
+                    throw new HttpRequestException(SR.net_http_max_redirects);
+                }
```

Comment:
Is this the behavior we decided on, throwing when the limit was reached rather than just returning whatever response we had when we hit that limit?  I don't have a strong opinion either way.

Reply:
I didn't change the logic here.  Not sure if we closed on this.  I filed #27285 to track.